### PR TITLE
User Profile: handle racing on creating new profile

### PR DESCRIPTION
**Pull request by ywangd**
This PR handles ensure multiple concurrent requests for creating new
profile for the same user get back the same profile.

This is made possible by having a predictable profile *base* uid for a user. 
The base uid is computed by hashing the username, realm or domain info.
The final uid is equal to "base uid + differentiator" where the differentiator
is a number that gets incremented if collision happens for the base uid.

The predictable base uid ensures only a single document is created for 
multiple concurrent ActivateProfile requests because ES does not allow
duplicate document IDs. 

For this approach to work correctly, the assumption is: At the time of
racing, realms and domains must be configured consistently for all 
nodes that can handle the ActivateProfile request. Inconsistent 
configurations can still lead to errors which will be addressed in 
separate future PRs.

### CODE REVIEW

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileDocument.java [hunk 1]
```diff
@@ -115,6 +119,25 @@ static ProfileDocument fromSubject(Subject subject) {
...
+                    digest.update(realmIdentifier.getName().getBytes(StandardCharsets.UTF_8));
+                }
+            });
+        } else {
+            digest.update(subject.getRealm().getType().getBytes(StandardCharsets.UTF_8));
+            if (false == subject.getRealm().isFileOrNative()) {
+                digest.update(subject.getRealm().getName().getBytes(StandardCharsets.UTF_8));
+            }
+        }
+        return "u_" + Base64.getUrlEncoder().withoutPadding().encodeToString(digest.digest());
```
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileDocument.java:138]: `withoutPadding` to avoid having `=` in the encoded string.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java [hunk 2]
```diff
@@ -344,18 +350,101 @@ private void createNewProfile(Subject subject, ActionListener<Profile> listener)
...
+            differentiatorNumber = Integer.parseInt(differentiator);
+        } catch (NumberFormatException e) {
+            listener.onFailure(new ElasticsearchException("profile uid [{}] differentiator is not a number", e, profileUid));
+            return;
+        }
+        // TODO: temporary to prevent infinite recursion
+        if (differentiatorNumber > 99) {
+            listener.onFailure(new ElasticsearchException("profile uid [{}] differentiator number is too high (>99)", profileUid));
+            return;
+        }
```
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java:410]: I don't think it could be a real scenario for having such high number of hash collision. But it is still nice to proactively prevent infinite recursion. This is just temporary. We can do better. One idea is to get all profile documents that has UID beginning with the clashed base uid and figure out the actual increment number instead of always incrementing by one in each recursion.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java [hunk 3]
```diff
@@ -344,18 +350,101 @@ private void createNewProfile(Subject subject, ActionListener<Profile> listener)
...
+                                // The profile document is a match, still need to update it with current auth info before return
+                                // listener.onResponse(versionedDocument.toProfile(Set.of()));
+                                updateProfileForActivate(subject, versionedDocument, listener);
+                            } else {
+                                // The profile document is NOT a match, this means either genuine hash collision or profile document
+                                // was manually updated, e.g. realm or user rename etc. So we attempt to differentiate from the
+                                // existing profile document by increase the differentiator number by 1.
+                                incrementDifferentiatorAndCreateNewProfile(subject, profileDocument, listener);
+                            }
+                        }, listener::onFailure));
```
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java:373]: As commented above, this `getVersionedDocument` call can be changed to get all documents with UID beginning with the baseUid for efficiency.

**Comment by albertzaharovits** [comment]
Let me try to describe the overall algorithm and summarize the changes proposed in this PR, with my own words, from a high level, in order to verify my understanding.

----

For every activate call, we search for the "corresponding" profile document. If it doesn't exist we create it.
The activate call contains user credentials that are authenticated by a realm which might or might not be part of a domain.

Profile documents are created for the username and its authenticating realm identifier (name and type).

The search for the corresponding profile takes into account the username and the domain realm identifiers. Therefore the search will find profiles that could have been created by other realms from the domain.

----

The issue now is that we want to support creating profiles concomitantly.
It is possible that multiple calls try to create profiles for different realms (but same username) which are part of the same domain.
This is a problem because it results in creating multiple profiles for the same domain which is an illegal state.

----

The proposed fix is that the document id of the profile document that is created be a function of the domain, instead of randomly generated.
This way, creating profiles for different realms under the same identical domain results in trying to create documents with the same id,only one of which will succeed.
However, there is also a chance of doc id collision with unrelated profile docs created before (the circumstances for this are a bit fuzzy).
To cater for this, there's an additional deterministic logic that profile creators use, again in order to ensure that for a single domain only one document is created.

----

As already discussed, I think this approach is OK. It's a quite intricate because we don't have a stable id for a profile document, but we still need to rely on the document id when creating profile documents.
I think the only way to improve on this is with stable 1:1 profile document ids to security domains, which probably has it's similar set of intricacies, but I think it's worth rehashing that in order to make sure we don't miss something obvious.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java [hunk 4]
```diff
@@ -344,18 +350,100 @@ private void createNewProfile(Subject subject, ActionListener<Profile> listener)
...
                         indexResponse.getPrimaryTerm(),
                         indexResponse.getSeqNo()
                     );
                     listener.onResponse(versionedDocument.toProfile(Set.of()));
-                }, listener::onFailure))
+                }, e -> {
+                    if (e instanceof VersionConflictEngineException) {
+                        // Document already exists with the specified ID, get the document with the ID
+                        // and check whether it is the right profile
+                        getVersionedDocument(profileDocument.uid(), ActionListener.wrap(versionedDocument -> {
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java:363]: Instead of the get doc here, it _might_ be simpler to reason about the whole logic if instead we redo the search.
The function that does the search (some internal function of `getVersionedDocument`), can take a parameter for the document id that it tries to use to create docs if the search returns empty. If the create doc fails, the method calls itself recursively with the incremented doc id.

Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java:363]: We do need improve the robustness for retrieving relevant documents after seeing a VersionConflict. Neither Search or Get can guarantee to return the necessary document in all cases (Get has a better chance because it is realtime but still not perfect). It's likely we need some backoff and retry strategy here. I will work on it with your suggestion in mind.

**Comment by ywangd** [comment]
@albertzaharovits Thanks a lot for checking this draft PR. I appreciate your time on it since it is rather involved.

> Let me try to describe the overall algorithm and summarize the changes proposed in this PR, with my own words, from a high level, in order to verify my understanding.

Yes your understanding is correct. There are certainly details that we can keep working on. At this stage, I just need to confirmation on the overall direction.

> It is possible that multiple calls try to create profiles for different realms (but same username) which are part of the same domain.

This is certainly true. But I think the most likely scenario is multiple requests trying to create profiles for the same username and *same* realm (thinking multiple components of a single Kibana page).

> However, there is also a chance of doc id collision with unrelated profile docs created before (the circumstances for this are a bit fuzzy).

I have two scenarios in mind for the collision: 
1. A genuine hash collision, i.e. somehow two different data actually end up having the same hash. This is very unlikley and almost impractical. 
2. Collision due to manual manipulation of profile document. For example, if a user changes its username (from `foo` to `bar`) after the profile is created, but still want to keep using the same profile. One way is to modify the profile document directly to match its new username so that the ActivateProfile API can still associate the user with the existing profile. Now, if a new user starts to use the system and the user happens to have the username `foo`, this new user will end up having the same document ID (hashing) as the old user. This is a very contrived example and I am not sure whether we would support manually updating profile document. But it is a possibility.

> As already discussed, I think this approach is OK. It's a quite intricate because we don't have a stable id for a profile document, but we still need to rely on the document id when creating profile documents.
I think the only way to improve on this is with stable 1:1 profile document ids to security domains, which probably has it's similar set of intricacies, but I think it's worth rehashing that in order to make sure we don't miss something obvious.

I am not sure why you said "we don't have a stable id ...". In most cases, we do have a stable ID. Collision is a very rare thing. Also stable ID is *not* import outside of handling racing for new profile document.

My read is that you are OK for me to develop this PR further into a formal one. I'll start working on it in next few days unless you have any objections. Thanks again!

In summary, both above scenarios are rather unlikely. But I think the code should handle them for good faith.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java [hunk 5]
```diff
@@ -357,18 +368,122 @@ private void createNewProfile(Subject subject, ActionListener<Profile> listener)
...
                         indexResponse.getPrimaryTerm(),
                         indexResponse.getSeqNo()
                     );
                     listener.onResponse(versionedDocument.toProfile(Set.of()));
-                }, listener::onFailure))
+                }, e -> {
+                    if (e instanceof VersionConflictEngineException) {
+                        // Document already exists with the specified ID, get the document with the ID
+                        // and check whether it is the right profile for the subject
+                        getOrCreateProfileWithBackoff(subject, profileDocument, DEFAULT_BACKOFF.iterator(), listener);
```
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java:381]: @albertzaharovits I decided to keep this to be a GET request instead of Search after some back and forth. Changing it to be a search adds complexity without real gain. Please let me explain:
* We cannot simply re-use the search that locates the existing profile. Because for an empty search result, we cannot tell whether the document is not yet available for search or does not match the search criteria. 
* Alternatively, we can augument the search with extra profile IDs, i.e. let it search for "username + realm/domain" as well "profile IDs" so that we can always get back all potentially relevant profiles. But with this approach, we cannot avoid checking whether "the document is the right one" locally which means we still need separate logic for search and local filtering.

Search is also slower and more expensive than GET. But racing on creating new profile should not happen very often, so it is probably not a big concern.

The only potential advantage that I can think of for using a search is that it has the potential to locate a profile that is being created from a different node with inconsistent configuration. For example, `node-1` has `realm1` without a domain, `node-2` has `realm1` and `realm2` with a domain. If they race to create a new profile, using search can potentially create/locate the same profile if there is an existing profile with clashing UID. The scenario is rather very unlikely. The behaviour seems to be smart. But I am not sure whether we should actively opt-in for it given (1) it's very rare and (2) we probably don't wanna cater for incosistent configuration at this point.

What do you think? I do have another version locally that uses search. So please let me know if you still prefer Search over GET.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java [hunk 6]
```diff
@@ -357,18 +368,122 @@ private void createNewProfile(Subject subject, ActionListener<Profile> listener)
...
+        try {
+            differentiator = Integer.parseInt(differentiatorString);
+        } catch (NumberFormatException e) {
+            throw new ElasticsearchException("profile uid [{}] differentiator is not a number", e, uid);
+        }
+        // Prevent infinite recursion. It is practically impossible to get this many clashes
+        if (differentiator > DIFFERENTIATOR_UPPER_LIMIT) {
+            listener.onFailure(new ElasticsearchException("profile uid [{}] differentiator number is too high", profileDocument.uid()));
+            return;
+        }
```
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java:455]: The upper limit of differentiator is 9 (0-based) which means it is not possible to have more than 10 profile documents sharing the same base UID. It is impractical to reach this number considering how clash can happen. Nevertheless, when it does happen, the API throw an error which is a non-ideal but defintive behaviour which in my opinion is acceptable for now.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java (hunk 5 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java:381]: Using search can also eliminate a few rounds of GET if the differentiator value is big (though with complexities in some other places). It can also potentially support a higher upper limit for the differentiator value (It is now 10 for the GET approach). So another "advantage" to consider.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java [hunk 7]
```diff
@@ -357,18 +367,101 @@ private void createNewProfile(Subject subject, ActionListener<Profile> listener)
...
+                        new ElasticsearchException("failed to retrieving profile [{}] after all retries", profileDocument.uid())
+                    );
+                }
+                return;
+            }
+            // Ownership check between the subject and the profile document
+            if (subject.canAccessResourcesOf(versionedDocument.doc.subject())) {
+                // The profile document can be accessed by the subject. It must have just got created by another thread, i.e. racing.
+                // Still need to update it with current auth info before return.
+                updateProfileForActivate(subject, versionedDocument, listener);
```
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java:421]: This `if` check basically reuses the existing logic from Authentication class so no additional logic is needed to filter profiles locally.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileDocument.java [hunk 8]
```diff
@@ -110,6 +126,29 @@ static ProfileDocument fromSubject(Subject subject) {
         );
     }
 
+    static String computeBaseUidForSubject(Subject subject) {
+        final MessageDigest digest = MessageDigests.md5();
```
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileDocument.java:130]: The length of final UID with `md5` digest is 20. It is 32 with `sha1` and 47 with `sha256`. I think `md5` should be sufficient for this use case. It is also available in BCFIPS lib.

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/profile/ProfileDomainIntegTests.java [hunk 9]
```diff
@@ -249,14 +259,137 @@ public void testGetProfileByAuthenticationWillNotCheckRealmNameForFileOrNativeRe
...
+            assertThat(profile1.uid(), equalTo(uid));
+            assertThat(profile1.user().username(), equalTo(username));
+        } else {
+            fail("Not all threads are ready after waiting");
+        }
+    }
+
+    public void testDifferentiator() {
+        String lastUid = null;
+        for (int i = 0; i < 10; i++) {
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/profile/ProfileDomainIntegTests.java:340]: No point in using a `10` size loop IMO, one activate after the `username` change is enough.

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/profile/ProfileDomainIntegTests.java [hunk 10]
```diff
@@ -249,14 +259,137 @@ public void testGetProfileByAuthenticationWillNotCheckRealmNameForFileOrNativeRe
...
+            }
+            // Manually update the username to create hash collision
+            final UpdateRequest updateRequest = client().prepareUpdate(SECURITY_PROFILE_ALIAS, "profile_" + currentUid).setDoc("""
+                {
+                  "user_profile": {
+                    "user": { "username": "should-not-be-found" }
+                  }
+                }
+                """, XContentType.JSON).setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE).request();
+            client().update(updateRequest).actionGet();
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/profile/ProfileDomainIntegTests.java:355]: I would also test that the profile with the name changed, can still be activated, ie it is not as if it was removed.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java [hunk 11]
```diff
@@ -128,7 +135,7 @@ public void activateProfile(Authentication authentication, ActionListener<Profil
 
         getVersionedDocument(subject, ActionListener.wrap(versionedDocument -> {
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java:136]: Should we rename this to "searchVersionedDocumentForSubject"? Particularly I would suggest that the method name contain "search" instead of "get".

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileDocument.java (hunk 8 above)
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileDocument.java:130]: I would use `sha256` for any new code, simply to avoid discussions like this one.

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/profile/ProfileDomainIntegTests.java [hunk 12]
```diff
@@ -249,14 +259,137 @@ public void testGetProfileByAuthenticationWillNotCheckRealmNameForFileOrNativeRe
         assertThat(future1.actionGet().uid(), equalTo(profile1.uid()));
     }
 
+    public void testConcurrentCreationOfNewProfiles() throws InterruptedException {
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/profile/ProfileDomainIntegTests.java:262]: Do you think a similar integ test for an already existing profile would be worthwhile?

Comment (albertzaharovits) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/profile/ProfileDomainIntegTests.java:262]: I mean test activating profiles concurrently when there are documents with a "collision" doc id.

Comment (albertzaharovits) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/profile/ProfileDomainIntegTests.java:262]: Because it is important that retrying to create the profile doc uses a deterministic new id as well (not just the first id is deterministic).

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java [hunk 13]
```diff
@@ -357,18 +367,101 @@ private void createNewProfile(Subject subject, ActionListener<Profile> listener)
...
+                        );
+                } else {
+                    // Retry has depleted. This can only happen when the document or the profile index itself gets deleted
+                    // in between requests.
+                    listener.onFailure(
+                        new ElasticsearchException("failed to retrieving profile [{}] after all retries", profileDocument.uid())
+                    );
+                }
+                return;
+            }
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java:416]: Do you think we could split this `getVersionedDocument` method in two parts here, and move the second part below this comment in the first's listener, such that the first part has the sole responsability of gettting the profile doc with retry.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java [hunk 14]
```diff
@@ -357,18 +367,101 @@ private void createNewProfile(Subject subject, ActionListener<Profile> listener)
...
+            // Ownership check between the subject and the profile document
+            if (subject.canAccessResourcesOf(versionedDocument.doc.subject())) {
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java:418]: Can we please have a unit test, or squeeze into an existing one, that the created profile docs pass the ownership check for the creator.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java (hunk 5 above)
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java:380]: Sounds good to me, thanks for considering it!

**Review by albertzaharovits — APPROVED** [review]
LGTM
We discussed and agreed on the algorithm beforehand.
I only had code style comments.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java [hunk 15]
```diff
@@ -357,18 +367,101 @@ private void createNewProfile(Subject subject, ActionListener<Profile> listener)
...
+                } else {
+                    // Retry has depleted. This can only happen when the document or the profile index itself gets deleted
+                    // in between requests.
+                    listener.onFailure(
+                        new ElasticsearchException("failed to retrieving profile [{}] after all retries", profileDocument.uid())
+                    );
+                }
+                return;
+            }
+            // Ownership check between the subject and the profile document
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java:417]: I would add some logging in the following if branches too.

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/profile/ProfileDomainIntegTests.java (hunk 9 above)
Comment (ywangd) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/profile/ProfileDomainIntegTests.java:340]: The loop is to test the erroring behaviour of `differentiator value is too high` which only happens when there are 10 collisions.

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/profile/ProfileDomainIntegTests.java [hunk 16]
```diff
@@ -249,14 +259,137 @@ public void testGetProfileByAuthenticationWillNotCheckRealmNameForFileOrNativeRe
...
+        // The profile index may or may not exist
+        if (randomBoolean()) {
+            indexDocument();
+        }
+
+        final String username = randomAlphaOfLengthBetween(5, 12);
+        final RealmDomain realmDomain = randomFrom(
```
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/profile/ProfileDomainIntegTests.java:269]: I randomized the domain using `AuthenticationTests#randomDomain(true)`. 

I intentionally make it possible to have `null` domain because it is perfect legit to run User Profile without any domain configuration and racing should be handled equally.

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/profile/ProfileDomainIntegTests.java (hunk 10 above)
Comment (ywangd) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/profile/ProfileDomainIntegTests.java:355]: Good point. It's added.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java (hunk 11 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java:136]: Good point. I updated as you suggested.

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/profile/ProfileDomainIntegTests.java (hunk 12 above)
Comment (ywangd) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/profile/ProfileDomainIntegTests.java:262]: Good point. I add a randomization to add an existing collision doc.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java (hunk 13 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java:416]: I tried to split them but at the end decided to keep it as is. The main reason is `getOrCreateProfileWithBackoff` is called in two places and it expects the same logic (the 2nd part) for handling an existing document. If we split the 2nd part out, we have to ensure the listener argument is wrapped properly to always ensure the 2nd part is included. It is doable. But I think it does not add too much value since the method isn't too long at this stage. And it has the future risk to change one call site without changing the other one (I actually made this mistake when refactoring it).

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java (hunk 15 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java:419]: Added.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java (hunk 14 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java:432]: This is sorta tested with `testConcurrentCreationOfNewProfiles` where this code is executed. But an explicit test is beneficial and I added one as `ProfileDomainIntegTests#testProfileDocumentPassCanAccessResourceCheck`. 

To ensure the search and the `canAccessResourcesOf` are always consistent, I also added this check right after the initial search of `activateProfile`. This adds an additional insurance that future changes should not break the functionally equality between the search and `canAccessResourcesOf` check.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java [hunk 17]
```diff
@@ -128,7 +135,7 @@ public void activateProfile(Authentication authentication, ActionListener<Profil
 
         getVersionedDocument(subject, ActionListener.wrap(versionedDocument -> {
             if (versionedDocument == null) {
-                createNewProfile(subject, listener);
+                createNewProfile(subject, ProfileDocument.fromSubject(subject), listener);
```
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java]: I refactored the method to take a `uid` instead of `ProfileDocument.fromSubject(subject)`.

**Comment by ywangd** [comment]
[Failure](https://gradle-enterprise.elastic.co/s/xtazpgttuy4w6/tests/:x-pack:plugin:ccr:qa:multi-cluster:follow-cluster/org.elasticsearch.xpack.ccr.AutoFollowIT/testAutoFollowPatterns?top-execution=1) is unrelated and tracked at https://github.com/elastic/elasticsearch/issues/84888

**Comment by ywangd** [comment]
@elasticmachine run elasticsearch-ci/part-2

**Comment by ywangd** [comment]
@elasticmachine run elasticsearch-ci/part-2

**Comment by ywangd** [comment]
@elasticmachine run elasticsearch-ci/part-2-fips
